### PR TITLE
add tests for keyboard shortcuts (#642)

### PR DESCRIPTION
**Pull request by jbhoosreddy**
It took me some time but I think I'm almost there. thanks to @jlongster and @jasonLaster's previous tests. I tried to replicate the original test to a good extent.

I would like to know if there is a better way `ensureThreadResumed` than what I did? Let me know what you think and I will make changes accordingly.

### CODE REVIEW

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  ok(findElement(dbg, "button", "pause", "active"), "Pause button disabled");
+  invokeInTab("firstCall");
+  yield waitForPaused(dbg);
+  ok(findElement(dbg, "button", "resume", "active"), "Resume button enabled");
+  pressKey(dbg, resumeKey);
+  yield waitForThreadEvents(dbg, "resumed");
+  yield waitForPaused(dbg);
+
+  // Evaluate script while debugger is paused
+  const result = yield evaluate(dbg, "1+1")
```
Comment (jlongster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I'm not sure I see the purpose of evaluating in this test. This doesn't really test anything related to keyboard shortcuts for stepping/resuming or the stepping buttons. This looks like it was ported over from the old test, but I think we can avoid the evaluate and adding the `evaluate` command for now.

Just trying to make the tests a little more focused than they were before.

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+];
+const resumeKey = "VK_F8";
+
+add_task(function*() {
+  const dbg = yield initDebugger(TAB_URL, ...sourceURLs);
+  const { selectors: { getBreakpoints, getBreakpoint }, getState } = dbg;
+  const sources = [];
+  for (let url of sourceURLs) {
+    sources.push(findSource(dbg, url));
+  }
```
Comment (jlongster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I think it's more descriptive and simple if you just do:

```js
const source1 = findSource(dbg, "switching-01.js");
const source2 = findSource(dbg, "switching-02.js");
```

Admittedly, it's not much more descriptive because the script names are just 01 and 02, but it's shorter code anyway. Most tests will only deal with 2 or 3 sources so it's not a big deal to do this for each, and it's easier to debug (you see `source1` and `source2` in scope instead of an array)

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  yield ensureAddBreakpoint(dbg, getState, getBreakpoint, sources[0].id, 5);
+  yield ensureAddBreakpoint(dbg, getState, getBreakpoint, sources[1].id, 6);
+  yield ensureAddBreakpoint(dbg, getState, getBreakpoint, sources[1].id, 7);
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I think it is okay to just add the breakpoint

`yield addBreakpoint()`

We're testing breakpoints in more depth in other spots. Also, the test will also be below when you expect to land in a certain spot.

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+const testResumeKey = Task.async(function* (dbg) {
+  info("Testing Resume Key");
+  ok(findElement(dbg, "button", "pause", "active"), "Pause button disabled");
+  invokeInTab("firstCall");
+  yield waitForPaused(dbg);
+  ok(findElement(dbg, "button", "resume", "active"), "Resume button enabled");
+  pressKey(dbg, resumeKey);
+  yield waitForThreadEvents(dbg, "resumed");
+  yield waitForPaused(dbg);
```
Comment (jlongster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: Very cool, thanks for porting this test. Awesome to have this stuff tested.

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  for (let url of sourceURLs) {
+    sources.push(findSource(dbg, url));
+  }
+
+  yield ensureAddBreakpoint(dbg, getState, getBreakpoint, sources[0].id, 5);
+  yield ensureAddBreakpoint(dbg, getState, getBreakpoint, sources[1].id, 6);
+  yield ensureAddBreakpoint(dbg, getState, getBreakpoint, sources[1].id, 7);
+  is(getBreakpoints(getState()).size, 3, "Three breakpoint exists");
+
+  yield testResumeButton(dbg);
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I think it would be better to expand some of this out into a couple lines

```js
invokeInTab("firstCall");
yield waitForPaused(dbg);
```

also, i think passing `.button.active` as a full selector is fine

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+});
+
+const ensureAddBreakpoint = Task.async(function* (dbg, getState, getBreakpoint,
+                                                  sourceId, line) {
+  yield addBreakpoint(dbg, sourceId, line);
+  ok(getBreakpoint(getState(), { sourceId, line }),
+    "Breakpoint has correct line");
+});
+
+const testResumeButton = Task.async(function* (dbg) {
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I'd prefer if this just tested the key shortcuts and not the buttons

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  // Evaluate script while debugger is paused
+  const result = yield evaluate(dbg, "1+1")
+    .then(response => response.result);
+  is(result, 2, "Evaluated script while debugger paused");
+  yield ensureThreadResumed(dbg);
+});
+
+const testResumeKey = Task.async(function* (dbg) {
+  info("Testing Resume Key");
+  ok(findElement(dbg, "button", "pause", "active"), "Pause button disabled");
```
Comment (jlongster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: You're passing 2 additional args to `findElement` but looks like the selector is `(button, state) => `.${button}.${state}`  ? It only takes 1 additional arg.

File: public/js/test/mochitest/browser_dbg_interrupts.js (hunk 3 above)
Comment (jlongster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: Personally I'd be fine with just doing `addBreakpoint` and not caring if the breakpoint slid or not. We know our test files, so it shouldn't slide. If it's mainly checking that it was successfully set, errors will propagate from `addBreakpoint` anyway.

**Comment by jlongster** [comment]
Several small nits but this generally is awesome :) Thanks for taking the time to do this, it helps *greatly*!

**Comment by jlongster** [comment]
Generally I wouldn't worry about porting old tests too carefully. They are generally bad, but they are good to look at to see specific edge cases to test. Feel free to write simple-ish tests for now and just look at the old tests to make sure there isn't something important we're not testing. Thanks for carefully looking through the old test.

File: public/js/test/mochitest/browser_dbg_interrupts.js (hunk 5 above)
Comment (jbhoosreddy) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: > I think it would be better to expand some of this out into a couple lines
Could you elaborate what you meant here?

Also do you mean this?
`findElement(dbg, "button", ".resume.active")`
and 
```
const selectors = {
  ...,
  button: selector => selector,
};
```

Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: actually, i think it would be better to have more elements like `findElement("resumeButton")`

**Comment by jbhoosreddy** [comment]
I modified the PR based on all the review comments.

File: public/js/components/RightSidebar.js [hunk 8]
```diff
@@ -88,15 +88,15 @@ const RightSidebar = React.createClass({
...
+    keyShortcuts.on("Shift+F11", this.stepOut);
   },
 
   componentWillUnmount() {
     const { keyShortcuts } = this.props;
     keyShortcuts.off("F8", this.resume);
     keyShortcuts.off("F10", this.stepOver);
     keyShortcuts.off("F11", this.stepIn);
-    keyShortcuts.off("F12", this.stepOut);
+    keyShortcuts.off("Shift+F11", this.stepOut);
```
Comment (jlongster) [inline public/js/components/RightSidebar.js]: Why this change? (may have miss the conversation) Is this what the old debugger does? It'd be good to keep the same shortcuts if possible, but there's a good reason that's ok.

File: public/js/test/mochitest/head.js [hunk 9]
```diff
@@ -225,8 +235,13 @@ function isVisibleWithin(outerEl, innerEl) {
...
   callStackHeader: ".call-stack-pane ._header",
   frame: (index) => `.frames ul li:nth-child(${index})`,
-  gutter: line => `.CodeMirror-code *:nth-child(${line}) .CodeMirror-linenumber`
-}
+  gutter: line => `.CodeMirror-code *:nth-child(${line}) .CodeMirror-linenumber`,
+  pauseKey: { keyCode: "VK_F8" },
+  resumeKey: { keyCode: "VK_F8" },
+  stepOverKey: { keyCode: "VK_F10" },
+  stepInKey: { keyCode: "VK_F11" },
+  stepOutKey: { keyCode: "VK_F12", modifiers: { shiftKey: true } }
```
Comment (jlongster) [inline public/js/test/mochitest/head.js]: Not sure these belong in `selectors`? These selectors are meant to be CSS selectors, and this just seems like a key mapping. You could pull this out into it's own object next to the `pressKey` function.

**Comment by jlongster** [comment]
New version looks great (w/a few nits), thanks!

**Comment by jbhoosreddy** [comment]
The change is because after I took out the button tests (as per Jason's
suggestion), there was not much my test was testing. So I decided to test
the different key shortcuts.

I could revert back if you guys think.

In either case, I'll work on these comments.

**Comment by jlongster** [comment]
> The change is because after I took out the button tests (as per Jason's suggestion), there was not much my test was testing. So I decided to test the different key shortcuts.

Oh testing all shortcuts is great :) I was talking about this change:

```
-    keyShortcuts.off("F12", this.stepOut);
+    keyShortcuts.off("Shift+F11", this.stepOut);
```

Which changed code, not a test, and changes the step out to be shift+F11 instead of F12.

**Comment by jlongster** [comment]
This is very close though. Thank you!

File: public/js/components/RightSidebar.js (hunk 8 above)
Comment (jbhoosreddy) [inline public/js/components/RightSidebar.js]: Yes. The old debugger has `Shift+F11` for `Step Out`. I wrote the code for the keyboard shortcuts.. and I think I chose `F12` by mistake. I will rectify that here.

File: public/js/test/mochitest/head.js (hunk 9 above)
Comment (jasonLaster) [inline public/js/test/mochitest/head.js]: agreed. This would make sense as a keyMappings dictionary.

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  yield testKey(dbg, "stepOutKey");
+});
+
+const testKey = Task.async(function* (dbg, key) {
+  info("Testing Key: " + key);
+  const { selectors: { getPause }, getState } = dbg;
+  yield waitForPaused(dbg);
+  ok(getPause(getState()), "Debugger is paused");
+  pressKey(dbg, key);
+});
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I don't think `testKey` is needed.

I'd prefer to invert the sequence. First press the key, and then check the pause location.

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  ok(!getPause(getState()), "Debugger is resumed");
+
+  yield testKey(dbg, "stepOutKey");
+});
+
+const testKey = Task.async(function* (dbg, key) {
+  info("Testing Key: " + key);
+  const { selectors: { getPause }, getState } = dbg;
+  yield waitForPaused(dbg);
+  ok(getPause(getState()), "Debugger is paused");
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: I have an `isPaused` function in my test that will be helpful here

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+  yield testKey(dbg, "stepInKey");
+  yield waitForThreadEvents(dbg, "resumed");
+  ok(!getPause(getState()), "Debugger is resumed");
+
+  yield testKey(dbg, "stepOutKey");
+});
+
+const testKey = Task.async(function* (dbg, key) {
+  info("Testing Key: " + key);
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: nit: I don't think we need the info, we'll see where it's failing with the right assertions

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  const source2 = findSource(dbg, "code-script-switching-02.js");
+
+  yield addBreakpoint(dbg, source1.id, 5);
+  yield addBreakpoint(dbg, source2.id, 6);
+  yield addBreakpoint(dbg, source1.id, 7);
+
+  yield testBreakOnNextState(dbg);
+
+  invokeInTab("firstCall");
+  yield testKey(dbg, "resumeKey");
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: There should be a `yield waitForPaused` here

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  yield waitForThreadEvents(dbg, "resumed");
+  ok(!getPause(getState()), "Debugger is resumed");
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: We should probably check the pause location instead of testing that at somepoint it was resumed:
```
assertPauseLocation(dbg, "code-script-switching-01.js", 12)
```

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  ok(getPause(getState()), "Debugger is paused");
+  pressKey(dbg, key);
+});
+
+const testBreakOnNextState = Task.async(function* (dbg) {
+  const { selectors: { getPause, getIsWaitingOnBreak }, getState } = dbg;
+  ok(!getPause(getState()), "Debugger is resumed");
+  pressKey(dbg, "pauseKey");
+  ok(getIsWaitingOnBreak(getState()), "Breaking on next execution");
+});
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: we don't need to test breaking on next here, so we just need to find a way to pause. Honestly, the easiest way might be to use the new `debugger-statements.html` i'm adding and go through the same steps as that test.

that way you don't need breakpoints or break on next

Comment (jbhoosreddy) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: a lot of these helpers are are there in your PR. I'll wait for that to get merged and then work on your comments. 👍

Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: cool. shouldn't be too long

**Comment by jasonLaster** [comment]
@jbhoosreddy #697 just landed

**Comment by jbhoosreddy** [comment]
My test is failing on Circle, but it looks fine on my end. I'm not too sure how to fix it though.

**Comment by jasonLaster** [comment]
Hmm, @jbhoosreddy have you rebased lately?

File: public/js/test/mochitest/browser_dbg_interrupts.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+  // step into
+  yield pressStepIn(dbg);
+  assertPausedLocation(dbg, "debugger-statements.html", 18);
+
+  // step over
+  yield pressStepOver(dbg);
+  assertPausedLocation(dbg, "debugger-statements.html", 20);
+
+  // step out
+  yield pressStepOut(dbg);
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: i believe my original test has these comments, but they're probably not necessary any more :)

File: public/js/test/mochitest/browser.ini [hunk 17]
```diff
@@ -26,3 +26,4 @@ support-files =
 [browser_dbg-chrome-debugging.js]
 [browser_dbg-iframes.js]
 [browser_dbg-debugger-buttons.js]
+[browser_dbg_interrupts.js]
```
Comment (jasonLaster) [inline public/js/test/mochitest/browser.ini]: we should probably rename this to `test-keyboard-shortcuts`

File: public/js/test/mochitest/browser_dbg_interrupts.js (hunk 16 above)
Comment (jbhoosreddy) [inline public/js/test/mochitest/browser_dbg_interrupts.js]: haha.. okay.. I can remove them.. :D

File: public/js/components/RightSidebar.js [hunk 18]
```diff
@@ -108,7 +112,7 @@ const RightSidebar = React.createClass({
     return [
       debugBtn(this.stepOver, "stepOver", className, "Step Over (F10)"),
       debugBtn(this.stepIn, "stepIn", className, "Step In (F11)"),
-      debugBtn(this.stepOut, "stepOut", className, "Step Out \u21E7 (F12)"),
+      debugBtn(this.stepOut, "stepOut", className, `Step Out (${shiftKey}F11)`),
```
Comment (jasonLaster) [inline public/js/components/RightSidebar.js]: I'm not sure how this will look with the parens

**Comment by jasonLaster** [comment]
👍 looks good minus some nits. I wonder if this is something to do with running on linux?

File: public/js/components/RightSidebar.js (hunk 18 above)
Comment (jbhoosreddy) [inline public/js/components/RightSidebar.js]: this is how it looks in the old debugger. I just replicated that.

**Comment by jasonLaster** [comment]
Just commenting for my own memory: @jbhoosreddy is looking into reproducing the linux failures, which look related to the key shortcuts library.

@jbhoosreddy If you'd like, it'd be okay if you removed those key shortcuts from this PR and then opened a new issue. The good thing with that is that fixing the key shortcuts library is a different kind of review

**Comment by jbhoosreddy** [comment]
@jasonLaster I updated the PR then. I commented out those lines and I will un-comment them once I fix it in a new PR.

**Comment by jasonLaster** [comment]
Thanks @jbhoosreddy! it always feels good to merge!
